modules/function/src/main/java/com/opengamma/strata/function/calculation/DefaultLookupRatesProvider.java

```diff
@@ -0,0 +1,280 @@
+import com.opengamma.strata.market.curve.InterpolatedNodalCurve;
+import com.opengamma.strata.market.id.SimpleCurveId;
+import com.opengamma.strata.market.key.IndexRateKey;
+import com.opengamma.strata.market.view.DiscountFactors;
+import com.opengamma.strata.market.view.DiscountFxForwardRates;
+import com.opengamma.strata.market.view.DiscountFxIndexRates;
+import com.opengamma.strata.market.view.FxForwardRates;
+import com.opengamma.strata.market.view.FxIndexRates;
+import com.opengamma.strata.market.view.IborIndexRates;
+import com.opengamma.strata.market.view.OvernightIndexRates;
+import com.opengamma.strata.market.view.PriceIndexValues;
+import com.opengamma.strata.pricer.rate.AbstractRatesProvider;
+import com.opengamma.strata.pricer.rate.RatesProvider;
+
+/**
+ * A rates provider based on a rates lookup.
+ * <p>
+ * This uses a {@link DefaultRatesMarketDataLookup} to provide a view on {@link MarketData}.
+ */
+@BeanDefinition(style = "light")
+final class DefaultLookupRatesProvider
+    extends AbstractRatesProvider
+    implements RatesProvider, ImmutableBean {
+
+  /**
+   * The lookup.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final DefaultRatesMarketDataLookup lookup;
+  /**
+   * The market data.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final MarketData marketData;
+  /**
+   * The FX rate provider.
+   */
+  private final FxRateProvider fxRateProvider;  // derived
+
+  //-------------------------------------------------------------------------
+  /**
+   * Obtains an instance based on a lookup and market data.
+   * <p>
+   * The lookup provides the mapping from currency to discount curve, and from
+   * index to forward curve. The curves are in the market data.
+   *
+   * @param lookup  the lookup
+   * @param marketData  the market data
+   * @return the rates provider
+   */
+  public static DefaultLookupRatesProvider of(DefaultRatesMarketDataLookup lookup, MarketData marketData) {
+    return new DefaultLookupRatesProvider(lookup, marketData);
+  }
+
+  @ImmutableConstructor
+  private DefaultLookupRatesProvider(DefaultRatesMarketDataLookup lookup, MarketData marketData) {
+    this.lookup = ArgChecker.notNull(lookup, "lookup");
+    this.marketData = ArgChecker.notNull(marketData, "marketData");
+    this.fxRateProvider = MarketDataFxRateProvider.of(marketData);
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public LocalDate getValuationDate() {
+    return marketData.getValuationDate();
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public <T> T data(MarketDataKey<T> key) {
+    return marketData.getValue(key);
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public Optional<Curve> findCurve(CurveName name) {
+    return Stream.concat(lookup.getDiscountCurves().values().stream(), lookup.getForwardCurves().values().stream())
```

Comment:
Currently there is not much lookup of curves by name. This can be more efficient, but only if we pass the curve identifier in sensitivities and other places. I haven't looked at that so far, as it hasn't been as necessary as I though it would be. Right now, this method is the best it can be in performance terms.